lucene/core/src/test/org/apache/lucene/index/TestExitableDirectoryReader.java

```diff
@@ -163,6 +163,17 @@ public void testExitableFilterTermsIndexReader() throws Exception {
     searcher.search(query, 10);
     reader.close();
 
+    // Set a fairly high timeout value (infinite) and expect the query to complete in that time frame.
+    // Not checking the validity of the result, but checking the sampling kicks in to reduce the number of timeout check
+    CountingQueryTimeout queryTimeout = new CountingQueryTimeout();
+    directoryReader = DirectoryReader.open(directory);
+    exitableDirectoryReader = new ExitableDirectoryReader(directoryReader, queryTimeout);
+    reader = new TestReader(getOnlyLeafReader(exitableDirectoryReader));
+    searcher = new IndexSearcher(reader);
+    searcher.search(query, 10);
+    reader.close();
+    assertEquals(3, queryTimeout.getShouldExitCallCount());
```

Comment:
Here we compare to the expected call count 3. Because 3 TermsEnum are created: it is a PrefixQuery and there is one TermsEnum created for AutomatonQuery.intersect() (the next call timeout check is skipped once), then 2 TermsEnum created for the 2 matching terms "one" and "ones").

Would it be clearer to have a separate test method?
We could index more docs, for example 50 with a prefix-generated term (e.g. "term"+increment).
There would be a PrefixQuery for "term", the same test code, and we would test the TIMEOUT_CHECK_SAMPLING:
- 1 TermsEnum for Automaton.intersect(), and the next calls would be sampled (50/TIMEOUT_CHECK_SAMPLING=3) => call count +4
- 1 TermsEnum for each enumerated term => call count +50
We could verify call count = 54 (instead of 101 without sampling)

Reply:
> Because 3 TermsEnum are created: it is a PrefixQuery and there is one TermsEnum created for AutomatonQuery.intersect() (the next call timeout check is skipped once), then 2 TermsEnum created for the 2 matching terms "one" and "ones").

Yes this is exactly what happened. Without the check the call count will be 4. 

------
I've tried to come up with a new test like what you suggested:
```
  /**
   * Tests time out check sampling of TermsEnum iterations
   * @throws Exception on error
   */
  public void testExitableTermsEnumSampleTimeoutCheck() throws Exception {
    try(Directory directory = newDirectory()) {
      try(IndexWriter writer = new IndexWriter(directory, newIndexWriterConfig(new MockAnalyzer(random())))) {
        for(int i = 0; i < 50; i++) {
          Document d1 = new Document();
          d1.add(newTextField("default", "term" + i, Field.Store.YES));
          writer.addDocument(d1);
        }

        writer.forceMerge(1);
        writer.commit();

        DirectoryReader directoryReader;
        DirectoryReader exitableDirectoryReader;
        IndexReader reader;
        IndexSearcher searcher;

        Query query = new PrefixQuery(new Term("default", "term"));

        // Set a fairly high timeout value (infinite) and expect the query to complete in that time frame.
        // Not checking the validity of the result, but checking the sampling kicks in to reduce the number of timeout check
        CountingQueryTimeout queryTimeout = new CountingQueryTimeout();
        directoryReader = DirectoryReader.open(directory);
        exitableDirectoryReader = new ExitableDirectoryReader(directoryReader, queryTimeout);
        reader = new TestReader(getOnlyLeafReader(exitableDirectoryReader));
        searcher = new IndexSearcher(reader);
        searcher.search(query, 300);
        reader.close();
        assertEquals(54, queryTimeout.getShouldExitCallCount());
      }
    }
  }
``` 
This test actually failed as the actual call count is 5 instead of 54. After some debugging it turns out that the threshold setting here https://github.com/apache/lucene-solr/blob/67ecd8ff9ad3640016424ded86bfaaadd815b96d/lucene/core/src/java/org/apache/lucene/search/MultiTermQueryConstantScoreWrapper.java#L118 also help reduce the query timeout check call count. If I were to change the threshold to 100, then the test will pass.

Reply:
This test is good. Can you actually add it to the PR? It should replace the lines added to testExitableFilterTermsIndexReader().

Your investigation reveals something I overlooked - the MultiTerm query is not rewritten to TermQueries above 16 matching terms.
- Just replace assertEquals(54, ...) by assertEquals(5, ...).
- Please add a comment above the assertEquals(5, ...) to explain it depends on ExitableDirectoryReader.ExitableTermsEnum.NUM_CALLS_PER_TIMEOUT_CHECK and MultiTermQueryConstantScoreWrapper.BOOLEAN_REWRITE_TERM_COUNT_THRESHOLD (so if someone changes one of these constants - unlikely -, she will know why this test fail and that it needs to be adjusted).
- I remove my remark about sampling ExitableTermsEnum constructor because we don't need it.

Reply:
Sounds good. I've added another commit to separate out the test, and adjusted assertion and added comments.